src/main/java/me/drepic/proton/ProtonManager.java

```diff
@@ -59,98 +52,21 @@ protected ProtonManager(String name, String[] groups, String host, String virtua
-            return;
-        }
-
-        String recipient = delivery.getProperties().getHeaders().get("recipient").toString();
-        String senderName = delivery.getProperties().getHeaders().get("x-senderName").toString();
-        UUID senderID = UUID.fromString(delivery.getProperties().getHeaders().get("x-senderID").toString());
-
-        if(senderID.equals(this.id) && recipient.isEmpty()){ //Implies this was a broadcast from us. Ignore
-            return;                                          //Conversely, we don't want to ignore messages we
-        }                                                    //purposefully sent ourself
-
-        if(!this.contextClassMap.containsKey(context)){ //Someone sent something to us that we're not listening for
-            Proton.pluginLogger().warning("Received message that has no registered handlers.");
-            return;
-        }
-
-        Class type = this.contextClassMap.get(context);
-        try{
-            Object body = gson.fromJson(msg, type);
-            MessageAttributes messageAttributes = new MessageAttributes(context.getNamespace(), context.getSubject(), senderName, senderID);
-            this.messageHandlers.get(context).forEach((biConsumer) -> {
-                try{
-                    biConsumer.accept(body, messageAttributes);
-                }catch(Exception e){
-                    e.printStackTrace();
-                }
-            });
-        }catch(Exception e){
-            e.printStackTrace();
-        }
-    }
-
-    private void bindRecipient(MessageContext context, String recipient) throws IOException {
-        Map<String, Object> headers = new HashMap<>();
-
-        headers.put("x-match", "all");
-        headers.put("recipient", recipient);
-        headers.put("messageContext", context.toContextString());
-
-        channel.queueBind(queueName, "proton.direct", "", headers);
     }
 
     /**
      * Registers the MessageContext with the direct exchange
      */
     private void registerMessageContext(MessageContext context) throws IOException {
 
-        bindRecipient(context, this.name);
+        this.bindRecipient(context, this.name);
 
         for (String group : this.groups) { //For each group this client belongs to, bind
             if(group.equals(this.name))continue; //prevent duplicate binding
-            bindRecipient(context, group);
+            this.bindRecipient(context, group);
         }
 
-        Map<String, Object> headers = new HashMap<>();
-        headers = new HashMap<>();
-        headers.put("x-match", "all");
-        headers.put("messageContext", context.toContextString());
-
-        channel.queueBind(queueName, "proton.broadcast", "", headers);
+        this.bindBroadcast(context);
     }
 
     private void internalSend(String namespace, String subject, Object data, Optional<String> recipient) {
```

Comment:
Don't need this method anymore. Have send() and broadcast() be abstract then `RabbitManager` (which probably should be named RabbitMQManager) can have a private internalSend() with the old contents of this method.

Reply:
This logic is common to both Redis and RabbitMQ though. Does it make sense to essentially bring it out and copy it twice.

Reply:
Well, this is the issue I see with the code structure as it is:

`public send()` and `public broadcast()` both call `private internalSend()`. `internalSend()` then has to figure out whether to call `protected send()` or `protected broadcast()`. To me this seems counterintuitive and I would refactor it in some way to remove the join which then splits off again.